include/dlaf/communication/sync/broadcast.h

```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
 }
 }
 
+/// Helper function for preparing a tile for sending.
+///
+/// Duplicates the tile to CPU memory if CUDA RDMA is not enabled for MPI.
+/// Returns the tile unmodified otherwise.
+template <Device D, typename T>
+auto prepareSendTile(hpx::shared_future<matrix::Tile<T, D>> tile) {
+  return matrix::duplicateIfNeeded<CommunicationDevice<D>::value>(std::move(tile));
+}
+
+/// Helper function for handling a tile after sending.
+///
+/// If CUDA RDMA is disabled, the tile returned from recvTile will always be on
+/// the CPU. This helper duplicates to the GPU if the first template parameter
+/// is a GPU device. The first template parameter must be given.
+template <Device D, typename T>
+auto handleRecvTile(hpx::future<matrix::Tile<T, CommunicationDevice<D>::value>> tile) {
+  return matrix::duplicateIfNeeded<D>(std::move(tile));
+}
+
 /// Task for broadcasting (send endpoint) a Tile in a direction over a CommunicatorGrid
-template <class T>
+template <typename T, Device D, template <class> class Future>
 void sendTile(hpx::future<common::PromiseGuard<comm::CommunicatorGrid>> mpi_task_chain, Coord rc_comm,
-              hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile) {
+              Future<matrix::Tile<T, D>> tile) {
```

Comment:
It might be ok to make it `shared_future`, but I'd prefer to avoid unnecessarily converting everything to a `shared_future` if not necessary.

Reply:
Can I ask you the rationale? Because usually I always tried to ensure const-correctness, while it seems there is a more important reason for not doing so.

Reply:
After discussing with @albestro I understood that I had read his question sloppily. I understood it to be about the change to have a generic `Future` type, rather than about the constness of `T`. It can and should indeed be `const` and I've just pushed a change which makes it `const`. Thanks @albestro for pointing this out!.